test/lit/passes/minimize-rec-groups.wast

```diff
@@ -0,0 +1,486 @@
+(module
+  (rec
+    ;; CHECK:      (rec
+    ;; CHECK-NEXT:  (type $b1 (struct (field (ref $a1))))
+
+    ;; CHECK:       (type $a1 (struct (field (ref $b1))))
+    (type $a1 (struct (field (ref $b1))))
+    (type $b1 (struct (field (ref $a1))))
+    ;; CHECK:      (rec
+    ;; CHECK-NEXT:  (type $2 (struct))
+
+    ;; CHECK:       (type $b2 (struct (field (ref $a2))))
+
+    ;; CHECK:       (type $a2 (struct (field (ref $b2))))
+    (type $a2 (struct (field (ref $b2))))
+    (type $b2 (struct (field (ref $a2))))
+  )
+
+  ;; CHECK:      (global $a1 (ref null $a1) (ref.null none))
+  (global $a1 (ref null $a1) (ref.null none))
+  ;; CHECK:      (global $a2 (ref null $a2) (ref.null none))
+  (global $a2 (ref null $a2) (ref.null none))
+)
+
+;; Now we can't differentiate by permutation because the subtyping constraint
+;; admits only one ordering.
+(module
+  (rec
+    ;; CHECK:      (rec
+    ;; CHECK-NEXT:  (type $a1 (sub (struct (field (ref $b1)))))
+    (type $a1 (sub (struct (field (ref $b1)))))
+    ;; CHECK:       (type $b1 (sub $a1 (struct (field (ref $b1)))))
+    (type $b1 (sub $a1 (struct (field (ref $b1)))))
+    ;; CHECK:      (rec
+    ;; CHECK-NEXT:  (type $2 (struct))
+
+    ;; CHECK:       (type $a2 (sub (struct (field (ref $b2)))))
+    (type $a2 (sub (struct (field (ref $b2)))))
+    ;; CHECK:       (type $b2 (sub $a2 (struct (field (ref $b2)))))
+    (type $b2 (sub $a2 (struct (field (ref $b2)))))
+  )
+
+  ;; CHECK:      (global $a1 (ref null $a1) (ref.null none))
+  (global $a1 (ref null $a1) (ref.null none))
+  ;; CHECK:      (global $a2 (ref null $a2) (ref.null none))
+  (global $a2 (ref null $a2) (ref.null none))
+)
+
+
+;; Now there are only two possible orderings admitted by the subtyping
+;; constraint.
+(module
+  (rec
+    ;; CHECK:      (rec
+    ;; CHECK-NEXT:  (type $a1 (sub (struct (field (ref $b1)))))
+    (type $a1 (sub (struct (field (ref $b1)))))
+    ;; CHECK:       (type $c1 (sub $a1 (struct (field (ref $b1)))))
+
+    ;; CHECK:       (type $b1 (sub $a1 (struct (field (ref $b1)) (field (ref $c1)))))
+    (type $b1 (sub $a1 (struct (field (ref $b1)) (ref $c1))))
+    (type $c1 (sub $a1 (struct (field (ref $b1)))))
+    ;; CHECK:      (rec
+    ;; CHECK-NEXT:  (type $a2 (sub (struct (field (ref $b2)))))
+    (type $a2 (sub (struct (field (ref $b2)))))
+    ;; CHECK:       (type $b2 (sub $a2 (struct (field (ref $b2)) (field (ref $c2)))))
+    (type $b2 (sub $a2 (struct (field (ref $b2)) (ref $c2))))
+    ;; CHECK:       (type $c2 (sub $a2 (struct (field (ref $b2)))))
+    (type $c2 (sub $a2 (struct (field (ref $b2)))))
+    ;; CHECK:      (rec
+    ;; CHECK-NEXT:  (type $6 (struct))
+
+    ;; CHECK:       (type $a3 (sub (struct (field (ref $b3)))))
+    (type $a3 (sub (struct (field (ref $b3)))))
+    ;; CHECK:       (type $c3 (sub $a3 (struct (field (ref $b3)))))
+
+    ;; CHECK:       (type $b3 (sub $a3 (struct (field (ref $b3)) (field (ref $c3)))))
```

Comment:
When we detect that we have two SCCs with the same shape, we form a nontrivial equivalence class for them to be members of. To do that, we find the canonical shape for the equivalence class and start generating permutations based on that. Calculating that canonical shape is what changes the order. In particular, the `$c` types are ordered before the `$b` types because they have fewer fields.